python/GafferTest/ThreadMonitorTest.py

```diff
@@ -0,0 +1,144 @@
+#        the distribution.
+#
+#      * Neither the name of John Haddon nor the names of
+#        any other contributors to this software may be used to endorse or
+#        promote products derived from this software without specific prior
+#        written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import threading
+import unittest
+
+import IECore
+
+import Gaffer
+import GafferTest
+
+class ThreadMonitorTest( GafferTest.TestCase ) :
+
+	def testConstruction( self ) :
+
+		monitor = Gaffer.ThreadMonitor()
+		self.assertEqual( monitor.allStatistics(), {} )
+		self.assertEqual( monitor.plugStatistics( Gaffer.IntPlug() ), {} )
+		self.assertEqual( monitor.combinedStatistics(), {} )
+
+	def testThisThreadId( self ) :
+
+		id = Gaffer.ThreadMonitor.thisThreadId()
+		self.assertEqual( id, Gaffer.ThreadMonitor.thisThreadId() )
+
+		ids = { id }
+		lock = threading.Lock()
+
+		def storeId() :
+			id = Gaffer.ThreadMonitor.thisThreadId()
+			self.assertEqual( id, Gaffer.ThreadMonitor.thisThreadId() )
+			with lock :
+				ids.add( id )
+
+		threads = []
+		for i in range( 0, 5 ) :
+			thread = threading.Thread( target = storeId )
+			threads.append( thread )
+			thread.start()
+
+		for thread in threads :
+			thread.join()
+
+		self.assertEqual( len( ids ), 6 )
+
+	def testMonitoring( self ) :
+
+		random = Gaffer.Random()
+		monitor = Gaffer.ThreadMonitor()
+
+		with monitor :
+			random["outFloat"].getValue()
+
+		self.assertEqual(
+			monitor.allStatistics(),
+			{
+				random["outFloat"] : {
+					monitor.thisThreadId() : 1
+				}
+			}
+		)
+		self.assertEqual(
+			monitor.plugStatistics( random["outFloat"] ),
+			{ monitor.thisThreadId() : 1 }
+		)
+		self.assertEqual(
+			monitor.combinedStatistics(),
+			{ monitor.thisThreadId() : 1 }
+		)
+
+		random["seedVariable"].setValue( "test" )
+		with monitor :
+			GafferTest.parallelGetValue( random["outFloat"], 100000, "test" )
+
+		s = monitor.plugStatistics( random["outFloat"] )
+		self.assertEqual( len( s ), IECore.tbb_global_control.active_value( IECore.tbb_global_control.parameter.max_allowed_parallelism ) )
+		self.assertEqual( sum( s.values() ), 100001 )
```

Comment:
This test would be a bit more convincing if there was some check of the distribution - fiddling with a Poisson distribution a little, if the selection of threads to run were independent, it should be basically guaranteed that every bucket get at least `0.5 * 100000 / max_allowed_parallelism` ... maybe lack of independence breaks this though? You've probably tested it in other ways so you know it works, I was just thinking that this would pass on [ 100001, 0, 0, 0, ... ]

Reply:
This does guarantee that each thread took at least 1 task (not 0), because otherwise there wouldn't be an entry for that thread at all (it's the processing of a task that creates the per-thread container in the first place).

I agree that we do want to test the quality of task distribution as well, but that's not really our concern when testing ThreadMonitor itself. I intend to use ThreadMonitor to provide such tests in ValuePlugTest and ComputeNodeTest in an upcoming TaskCollaboration pull request.

Reply:
I did _completely_ mess up the process mask test though - that fixed in https://github.com/GafferHQ/gaffer/pull/5425/commits/f7341df9f203cbaccd56994fa8789929f72ed338.